# Desafio: Explorando IA Generativa em um Pipeline de ETL com Python

Código feito no **Jupter noteboon** ou **Google Colab**.

1. Foi usada uma planilha simples, em formato CSV ('SDW2023.csv'), com uma lista de IDs de usuário do banco, segue exemplo a baixo:

<table>
    <thead>
        <tr align="center">
            <th>UserID</th>
        </tr>
    </thead>
    <thead>
        <tr>
            <td align="center"> 1 </td>
        </tr>
        <tr>
            <td align="center"> 2 </td>
        </tr>
        <tr>
            <td align="center"> 3 </td>
        </tr>
        <tr>

2. Consumir o endpoint GET https://sdw-2023-prd.up.railway.app/users/{id} (API da Santander Dev Week 2023) para obter os dados de cada cliente.

In [7]:
# Repositório da API: https://github.com/digitalinnovationone/santander-dev-week-2023-api
# Irei renomear o nome da URL para não ficar grande no código.
sdw2023_api_url = 'https://sdw-2023-prd.up.railway.app'

# Extract

3. Extraindo a lista de IDs de usuário a partir do arquivo CSV. Para cada ID, iremos fazer uma requisição GET para obter os dados do usuário correspondente.

In [8]:
# Importanta instalar o Pandas através do terminal caso não esteja instalado, comando: pip install pandas
# importando o pandas e executando o código para buscar o arquivo.
import pandas as pd

df = pd.read_csv('SDW2023.csv')
user_ids = df['UserID'].tolist()
print(user_ids)

[641, 642, 643]


### Create new user

4. Lembrando que lá no site da API do Santander é preciso você ir na opção "POST" e criar um usuário. Apartir do momento que o usuário for criado o sistema irá gerar uma "ID" automatica e é essa "ID" que deve ser preenchida no seu Dataframe.

No meu caso foi gerado as IDs 641, 642 e 643.


In [9]:
# Vamos rodar o comando abaixo para verificar as primeiras linhas do nosso Dataframe:

df.head()

UserID
0     641
1     642
2     643

In [10]:
# Importando o request e json para se conectar no Site de teste da API do Santander
import requests
import json

def get_user(id):
  response = requests.get(f'{sdw2023_api_url}/users/{id}')
  return response.json() if response.status_code == 200 else None

users = [user for id in user_ids if (user := get_user(id)) is not None]
print(json.dumps(users, indent=2))

[
  {
    "id": 641,
    "name": "Cunha",
    "account": {
      "id": 660,
      "number": "01.097922-2",
      "agency": "2022",
      "balance": 0.0,
      "limit": 0.0
    },
    "card": {
      "id": 615,
      "number": "xxxx xxxx xxxx 2022",
      "limit": 0.0
    },
    "features": [
      {
        "id": 233,
        "icon": "string",
        "description": "string"
      }
    ],
    "news": [
      {
        "id": 1862,
        "icon": "string",
        "description": "string"
      }
    ]
  },
  {
    "id": 642,
    "name": "Wenderes",
    "account": {
      "id": 661,
      "number": "01.097933-3",
      "agency": "2033",
      "balance": 0.0,
      "limit": 0.0
    },
    "card": {
      "id": 616,
      "number": "xxxx xxxx xxxx 2033",
      "limit": 0.0
    },
    "features": [
      {
        "id": 234,
        "icon": "string",
        "description": "string"
      }
    ],
    "news": [
      {
        "id": 1863,
        "icon": "string",
        "description": "st

# Transform
Essa é a parte que iremos tratar e transformar os dados!

5. Iremos utilizar a API do OpenAI GPT-3.5 para gerar uma mensagem de marketing personalizada para cada usuário.

Lembrando que precisa instalar a biblioteca da API da OpenAI do ChatGPT antes de importala.

In [11]:
# Documentação Oficial da API OpenAI: https://platform.openai.com/docs/api-reference/introduction
# Link direto: https://platform.openai.com/account/api-keys
# Substitua o texto por sua API Key da OpenAI, ela será salva como uma variável de ambiente.

openai_api_key = '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'

In [ ]:
import openai

openai.api_key = openai_api_key


def generate_ai_news(user):
  completion = openai.ChatCompletion.create(
    model="gpt-3.5-turbo",
    messages=[
      {
          "role": "system",
          "content": "Você é um especialista em markting bancário e investimentos."
      },
      {
          "role": "user",
          "content": f"Crie uma mensagem para {user['name']} sobre a importância de saber usar o dinheiro (máximo de 100 caracteres)"
      }
    ]
  )
  return completion.choices[0].message.content.strip('\"')  

for user in users:
  news = generate_ai_news(user)
  print(news)
    
  user['news'].append({
      "icon": "https://digitalinnovationone.github.io/santander-dev-week-2023-api/icons/credit.svg",
      "description": news
})  

# Load

6. Atualizanda a lista de "news" de cada usuário na API com a nova mensagem gerada.

In [ ]:
def update_user(user):
  response = requests.put(f"{sdw2023_api_url}/users/{user['id']}", json=user)
  return True if response.status_code == 200 else False

for user in users:
  success = update_user(user)
  print(f"User {user['name']} updated? {success}!")